hypothesis-python/src/hypothesis/searchstrategy/misc.py

```diff
@@ -112,18 +112,23 @@ def check_index(i):
             if check_index(i):
                 return self.elements[i]
 
+        # If we've tried all the possible elements, give up now.
+        max_good_indices = len(self.elements) - len(known_bad_indices)
+        if not max_good_indices:
+            return filter_not_satisfied
+
         # Figure out the bit-length of the index that we will write back after
         # choosing an allowed element.
         write_length = bit_length(len(self.elements))
 
         # Impose an arbitrary cutoff to prevent us from wasting too much time
         # on very large element lists.
         cutoff = 10000
+        max_good_indices = min(max_good_indices, cutoff)
```

Comment:
The one change I made here was to once again apply `cutoff` to `max_good_indices`.

This isn't necessary for correctness, but it greatly increases the probability of choosing an in-bounds speculative index when there are many elements.

Reply:
(Maybe I should try to think of a way to test this in a follow-up PR.)

Reply:
I had thought about moving from a cutoff for long lists to a "choose random 1K-element slice, then select from those valid indices if any" - a bit more bookkeeping, but it scales more evenly to longer lists and we can tune the size of the slice more to balance typical runtime against chance to succeed.